File: app/src/main/java/com/gmail/fattazzo/meteo/preferences/iconthemepicker/IconThemePickerPreference.java
package com.gmail.fattazzo.meteo.preferences.iconthemepicker;

import android.app.AlertDialog.Builder;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.content.res.Resources;
import android.preference.ListPreference;
import android.preference.PreferenceManager;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.gmail.fattazzo.meteo.R;
import com.gmail.fattazzo.meteo.utils.icons.WeatherIconsFactory;
import com.gmail.fattazzo.meteo.utils.icons.retirever.IconsRetriever;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author fattazzo
 * <p/>
 * date: 23/02/16
 */
public class IconThemePickerPreference extends ListPreference {

    private static final String TAG = "IconThemePreference";

    private Context context;
    private ImageView icon1;
    private ImageView icon2;
    private ImageView icon3;

    private CharSequence[] iconName;
    private List<IconThemeItem> icons;
    private SharedPreferences preferences;
    private Resources resources;
    private String selectedIconPath, defaultIconFile;
    private TextView summary;

    public IconThemePickerPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.context = context;

        resources = context.getResources();
        preferences = PreferenceManager.getDefaultSharedPreferences(context);

        defaultIconFile = "meteo_trentino";
    }

    private String getEntry(String value) {
        String[] entries = resources.getStringArray(R.array.iconThemeName);
        String[] values = resources.getStringArray(R.array.iconThemeKey);
        int index = Arrays.asList(values).indexOf(value);
        try {
            return entries[index];
        } catch (Exception e) {
            return entries[0];
        }

    }

    @Override
    protected void onBindView(View view) {
        super.onBindView(view);

        selectedIconPath = preferences.getString(
                context.getResources().getString(R.string.pref_key_iconsTheme), defaultIconFile);

        icon1 = (ImageView) view.findViewById(R.id.iconSelected1);
        icon2 = (ImageView) view.findViewById(R.id.iconSelected2);
        icon3 = (ImageView) view.findViewById(R.id.iconSelected3);
        updateIcon();

        summary = (TextView) view.findViewById(R.id.icon_theme_pref_summary);
        summary.setText(getEntry(selectedIconPath));

    }

    @Override
    protected void onDialogClosed(boolean positiveResult) {
        super.onDialogClosed(positiveResult);

        if (icons != null) {
            for (int i = 0; i < iconName.length; i++) {
                IconThemeItem item = icons.get(i);
                if (item.isChecked()) {

                    Editor editor = preferences.edit();
                    editor.putString(
                            context.getResources().getString(R.string.pref_key_iconsTheme),
                            item.getPath());
                    editor.apply();

                    selectedIconPath = item.getPath();
                    updateIcon();

                    summary.setText(item.getName());

                    break;
                }
            }
        }

    }

    @Override
    protected void onPrepareDialogBuilder(Builder builder) {

        builder.setNegativeButton("Cancel", null);
        builder.setPositiveButton(null, null);

        iconName = getEntries();
        CharSequence[] iconPath = getEntryValues();

        if (iconName == null || iconPath == null
                || iconName.length != iconPath.length) {
            throw new IllegalStateException(
                    "ListPreference requires an entries array "
                            + "and an entryValues array which are both the same length");
        }

        String selectedIcon = preferences.getString(
                context.getResources().getString(R.string.pref_key_iconsTheme),
                "meteo_trentino");

        icons = new ArrayList<>();

        for (int i = 0; i < iconName.length; i++) {
            boolean isSelected = selectedIcon.equals(iconPath[i]);
            IconThemeItem item = new IconThemeItem(iconName[i], iconPath[i], isSelected);
            icons.add(item);
        }

        IconThemeListPreferenceAdapter adapter = new IconThemeListPreferenceAdapter(
                context, R.layout.icon_theme_item_picker, icons);
        adapter.setOnThemeSelectedClosure(paramObject -> getDialog().dismiss());
        builder.setAdapter(adapter, null);

    }

    private void updateIcon() {
        IconsRetriever iconsRetriever = WeatherIconsFactory.INSTANCE.getIconsRetriever(selectedIconPath);

        icon1.setImageResource(iconsRetriever.getIconDemo1());
        icon1.setTag(selectedIconPath);

        icon2.setImageResource(iconsRetriever.getIconDemo2());
        icon2.setTag(selectedIconPath);

        icon3.setImageResource(iconsRetriever.getIconDemo3());
        icon3.setTag(selectedIconPath);
    }

}

File: app/src/main/java/com/gmail/fattazzo/meteo/preferences/view/iconthemepicker/IconThemePickerPreference.kt
/*
 * Project: meteo
 * File: IconThemePickerPreference.kt
 *
 * Created by fattazzo
 * Copyright © 2019 Gianluca Fattarsi. All rights reserved.
 *
 * MIT License
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.gmail.fattazzo.meteo.preferences.view.iconthemepicker

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Resources
import android.util.AttributeSet
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.preference.ListPreference
import androidx.preference.PreferenceManager
import androidx.preference.PreferenceViewHolder
import com.gmail.fattazzo.meteo.R
import com.gmail.fattazzo.meteo.utils.icons.WeatherIconsFactory
import java.util.*

/**
 * @author fattazzo
 *
 *
 * date: 23/02/16
 */
class IconThemePickerPreference(context: Context, attrs: AttributeSet) :
    ListPreference(context, attrs) {
    private val defaultIconFile: String = "meteo_trentino"

    private var icon1: ImageView? = null
    private var icon2: ImageView? = null
    private var icon3: ImageView? = null

    private var iconName: Array<CharSequence>? = null
    private var icons: MutableList<IconThemeItem>? = null
    private val preferences: SharedPreferences =
        PreferenceManager.getDefaultSharedPreferences(context)
    private val resources: Resources = context.resources
    private var selectedIconPath: String =
        preferences.getString(context.resources.getString(R.string.pref_key_iconsTheme), null)
            ?: defaultIconFile

    private var summary: TextView? = null

    private fun getEntry(value: String): String {
        val entries = resources.getStringArray(R.array.iconThemeName)
        val values = resources.getStringArray(R.array.iconThemeKey)
        val index = listOf(*values).indexOf(value)
        return try {
            entries[index]
        } catch (e: Exception) {
            entries[0]
        }

    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)

        val view = holder.itemView

        view.setPadding(50, 50, 50, 50)

        val iconIV = view.findViewById<ImageView>(R.id.icon)
        iconIV.setImageDrawable(icon)

        //selectedIconPath = preferences.getString(context.resources.getString(R.string.pref_key_iconsTheme), defaultIconFile)

        icon1 = view.findViewById(R.id.iconSelected1)
        icon2 = view.findViewById(R.id.iconSelected2)
        icon3 = view.findViewById(R.id.iconSelected3)
        updateIcon()

        summary = view.findViewById(R.id.icon_theme_pref_summary)
        summary!!.text = getEntry(selectedIconPath)
    }

    override fun onClick() {
        val builder = AlertDialog.Builder(context)
        builder.setNegativeButton("Cancel", null)
        builder.setPositiveButton("Ok") { dialog, _ ->
            if (icons != null) {
                for (i in iconName!!.indices) {
                    val item = icons!![i]
                    if (item.isChecked) {

                        persistString(item.path)
                        callChangeListener(item.path)

                        selectedIconPath = item.path
                        updateIcon()

                        summary!!.text = item.name

                        break
                    }
                }
            }
            dialog.dismiss()
        }

        iconName = entries
        val iconPath = entryValues

        check(
            !(iconName == null || iconPath == null
                    || iconName!!.size != iconPath.size)
        ) { "ListPreference requires an entries array " + "and an entryValues array which are both the same length" }

        val selectedIcon = getPersistedString("meteo_trentino")

        /**
         * String selectedIcon = preferences.getString(
         * context.getResources().getString(R.string.pref_key_iconsTheme),
         * "meteo_trentino");
         */

        icons = ArrayList()

        for (i in iconName!!.indices) {
            val isSelected = selectedIcon == iconPath[i]
            val item = IconThemeItem(
                iconName!![i],
                iconPath[i],
                isSelected
            )
            icons!!.add(item)
        }

        val adapter =
            IconThemeListPreferenceAdapter(
                context, R.layout.icon_theme_item_picker, icons!!
            )
        builder.setAdapter(adapter, null)
        builder.create().show()
    }

    private fun updateIcon() {
        val iconsRetriever = WeatherIconsFactory.getIconsRetriever(selectedIconPath)

        icon1!!.setImageResource(iconsRetriever.getIconDemo1())
        icon1!!.tag = selectedIconPath

        icon2!!.setImageResource(iconsRetriever.getIconDemo2())
        icon2!!.tag = selectedIconPath

        icon3!!.setImageResource(iconsRetriever.getIconDemo3())
        icon3!!.tag = selectedIconPath
    }
}